### ISSUE TITLE
NCudaLib::TOutOfMemoryError

### ISSUE THREAD

**Issue by david-waterworth** [catboost/catboost#583]
Problem: 

When attempting to train on my GPU, catboost is trying to allocate 100% of the GPU memory, despite other processes using some of it.Is there some way of preventing this or is it just that the size of the problem is larger than my GPU memory (I don't have this issue on my home machine which has 2 x 1080Ti's)

> (NCudaLib::TOutOfMemoryError) catboost/cuda/cuda_lib/memory_pool/stack_like_memory_pool.h:305: Error: Out of memory. Requested 11448 MB; Free 7790.003857 MB
> uncaught exception:
>     address -> 0x106f4f808
>     what() -> "catboost/cuda/cuda_lib/memory_pool/stack_like_memory_pool.h:305: Error: Out of memory. Requested 11448 MB; Free 7790.003857 MB"
>     type -> NCudaLib::TOutOfMemoryError
> Aborted (core dumped)

I'm using CatBoostRegressor

catboost version: 0.11.1 (python3)
Operating System: Ubuntu 18.04
GPU: Nvidia 2080 Ti

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.72       Driver Version: 410.72       CUDA Version: 10.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce RTX 208...  Off  | 00000000:01:00.0  On |                  N/A |
| 35%   38C    P2    53W / 260W |   2471MiB / 10986MiB |      3%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1556      C   /usr/lib/libreoffice/program/soffice.bin     171MiB |
|    0      2292      G   /usr/lib/xorg/Xorg                            40MiB |
|    0      2338      G   /usr/bin/gnome-shell                          58MiB |
|    0      2767      G   /usr/lib/xorg/Xorg                           597MiB |
|    0      2899      G   /usr/bin/gnome-shell                         398MiB |
|    0      6016      G   ...-token=8BC91079EDD400728C4BBC69E18389F1   307MiB |
|    0      6484      G   ...uest-channel-token=12750394730543361840   892MiB |
+-----------------------------------------------------------------------------+

```

**Comment by snovik75** [comment]
I have the same issue with 0.15.1 also on linux. It is really annoy to see

Out of memory on GPU

**Comment by dimitar-petrov** [comment]
I am facing similar issue:
```
Application terminated with error: ??+0 (0x7F8CCC5213EE)
??+0 (0x7F8CCC51F37E)
??+0 (0x7F8CCC51E035)
??+0 (0x7F8CCC51EAA8)
??+0 (0x7F8CCB848733)
??+0 (0x7F8CCB8485E7)
??+0 (0x7F8D3585BA9D)
clone+67 (0x7F8D3578BB23)

(NCudaLib::TOutOfMemoryError) catboost/cuda/cuda_lib/memory_pool/stack_like_memory_pool.h:305: Error: Out of memory. Requested 30559.5 MB; Free 7577.129639 MB

```

Is there a work around to limit the memory usage.

**Comment by annaveronika** [comment]
We've finally published docs what to do if you have out of memory error on gpu:
https://catboost.ai/docs/concepts/faq.html#overcome-out-of-memory

**Comment by david-waterworth** [comment]
@annaveronika this doesn't fix the issue I originally raised. There was talk that you'd add the ability to manually (or a flag to automatically) free GPU memory due to the cost of the cuda allocator (i.e. you don't want to allocated and free the memory on each train by default?

The issue I have is if I run catboost in a jupyter notebook each kernel appears to attempt to allocate around 80% of the available GPU memory. Normal python scripts return this memory when they exit but jupyter kernels need to be manually terminated so I find it's not really feasible to use catboost with jupyter as I'm constantly killing kernels (and sometimes they panic themselves).

**Comment by CallShaul** [comment]
Having same issue, while using a 24GB memory card:

(class NCudaLib::TOutOfMemoryError) c:/program files (x86)/go agent/pipelines/buildmaster/catboost.git/catboost/cuda/cuda_lib/memory_pool/stack_like_memory_pool.h:303: Error: Out of memory. Requested 6556.213867 MB; Free 1837.010351 MB

![image](https://user-images.githubusercontent.com/52006780/103289238-59f26000-49ef-11eb-85c5-c039cae94916.png)

**Comment by FAFUuser** [comment]
> 我遇到了类似的问题：
> 
> ```
> Application terminated with error: ??+0 (0x7F8CCC5213EE)
> ??+0 (0x7F8CCC51F37E)
> ??+0 (0x7F8CCC51E035)
> ??+0 (0x7F8CCC51EAA8)
> ??+0 (0x7F8CCB848733)
> ??+0 (0x7F8CCB8485E7)
> ??+0 (0x7F8D3585BA9D)
> clone+67 (0x7F8D3578BB23)
> 
> (NCudaLib::TOutOfMemoryError) catboost/cuda/cuda_lib/memory_pool/stack_like_memory_pool.h:305: Error: Out of memory. Requested 30559.5 MB; Free 7577.129639 MB
> ```
> 
> 是否有解决方法来限制内存使用。

Hellow！
I have encountered such a problem recently. How did you solve it? Thank you very much！

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
hi @FAFUuser !
assuming your dataset contains only numeric features, here are very approximate estimates for required gpu ram (in bytes) 
regression and classification -- (number of samples)*(number of features)
ranking -- (number of samples)*(number of features)*(average number of samples in a ranking group)
since catboost may allocate temporary buffers in gpu ram, i would multiply this by 1.2


to be more specific, could you please post here an end-to-end reproducer, the details about your hardware (type and number of gpus, etc.), the vesrion of catboost you are using...

**Comment by FAFUuser** [comment]
> hi @FAFUuser ! assuming your dataset contains only numeric features, here are very approximate estimates for required gpu ram (in bytes) regression and classification -- (number of samples)*(number of features) ranking -- (number of samples)*(number of features)*(average number of samples in a ranking group) since catboost may allocate temporary buffers in gpu ram, i would multiply this by 1.2
> 
> to be more specific, could you please post here an end-to-end reproducer, the details about your hardware (type and number of gpus, etc.), the vesrion of catboost you are using...

Hi @Evgueni-Petrov-aka-espetrov ！
First of all, thank you very much for your reply! My dataset contains only numerical features, but the data volume is indeed relatively large at 85GB
My GPU type is
NVIDIA GeForce RTX 3090.
Dedicated GPU memory 24.0 GB.
Shared GPU memory 75.0 GB.
GPU memory 99.0 GB.
CatBoost version is: 1.0.6
Thank you again for all your help and I look forward to hearing from you.

**Comment by FAFUuser** [comment]
> hi @FAFUuser ! assuming your dataset contains only numeric features, here are very approximate estimates for required gpu ram (in bytes) regression and classification -- (number of samples)*(number of features) ranking -- (number of samples)*(number of features)*(average number of samples in a ranking group) since catboost may allocate temporary buffers in gpu ram, i would multiply this by 1.2
> 
> to be more specific, could you please post here an end-to-end reproducer, the details about your hardware (type and number of gpus, etc.), the vesrion of catboost you are using...

Meanwhile，My model parameters are as follows：
   catmodel = catr.CatBoostRegressor(depth=6, iterations=200, learning_rate=0.1,
                                                             verbose=False, task_type="GPU",
                                                              used_ram_limit=None,gpu_ram_part=0.95, loss_function='RMSE',)

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
if it's just training a model, then you may down-sample the dataset so that it fits into gpu memory

if it's some sort of hyperopt, then please consider switching to CatBoost.randomized_search https://catboost.ai/en/docs/concepts/python-reference_catboost_randomized_search
hyperopt in libraries like sklearn may run multiple CatBoost trainings in parallel, which may lead to out-of-memory on gpu

**Comment by FAFUuser** [comment]
> if it's just training a model, then you may down-sample the dataset so that it fits into gpu memory
> 
> if it's some sort of hyperopt, then please consider switching to CatBoost.randomized_search in addition to down-sampling https://catboost.ai/en/docs/concepts/python-reference_catboost_randomized_search hyperopt in libraries like sklearn may run multiple CatBoost trainings in parallel, which may lead to out-of-memory on gpu

Thank you very much for your help and I will try the solution strategy you have provided. Please allow me to ask you a question again. In the program I am running, Python is giving me the following error message，ERROR: out of memory. requested 1795MB; free 595 MB. I am confused by the fact that my GPU has much more than 595MB of memory and that the dataset used is more than 1795MB.I look forward to hearing from you, good luck with your work and have a great life!

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
gpu memory is allocated step-by-step
this message means that catboost wanted to allocated 1795Mb and only 595Mb was available by that moment
it is not related to the size of the gpu memory

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
many thanks for your kind wishes :)

**Comment by FAFUuser** [comment]
I see, thank you and good luck！:)

**Comment by kizill** [comment]
Closing, as a fixed duplicate of #619

**Comment by Bwahharharrr** [comment]
This thread was closed as a duplicate issue 619 however 619 say its been fixed, and 619 says it was fixed 4 years before this thread was started. Another 2 years later and i'm also suffering regular training crashes due to the same issue.

PyInit__catboost+10380359 (0x7FFB65585A6C)
PyInit__catboost+42712873 (0x7FFB6745B54E)
PyInit__catboost+42707002 (0x7FFB67459E5F)
PyInit__catboost+42714750 (0x7FFB6745BCA3)
PyInit__catboost+42696484 (0x7FFB67457549)
PyInit__catboost+42698024 (0x7FFB67457B4D)
PyInit__catboost+5765476 (0x7FFB6511EF89)
PyInit__catboost+47359510 (0x7FFB678C9C3B)
BaseThreadInitThunk+23 (0x7FFC8894E8D7)
RtlUserThreadStart+44 (0x7FFC8A93C34C)

(class NCudaLib::TOutOfMemoryError) catboost/cuda/cuda_lib/memory_pool/stack_like_memory_pool.h:303: Error: Out of memory. Requested 4935.9375 MB; Free 4282.649804 MB
